File: R/old/get_cm360_files.R
get_cm360_files <- function(profile_id, report_id, credentials = NULL) {
  pacman::p_load(tidyverse, reticulate)
  
  # If credentials are not provided, refresh them
  if (is.null(credentials)) {
    credentials <- gcp_reauth()
  }
  
  # Import Google API client
  googleapiclient <- import("googleapiclient.discovery")
  cm360 <- googleapiclient$build("dfareporting", "v4", credentials = credentials)
  
  # List all files for the report
  files <- cm360$reports()$files()$list(profileId = profile_id, reportId = report_id)$execute()
  
  # If no files exist, return an empty data frame
  if (is.null(files$items) || length(files$items) == 0) {
    warning("No files found for this report.")
    return(tibble(
      file_id = character(), report_id = character(), etag = character(), status = character(),
      file_name = character(), format = character(), last_modified = POSIXct(),
      start_date = character(), end_date = character(), browser_url = character(),
      api_url = character()
    ))
  }
  
  # Convert nested list to a data frame
  files_df <- files$items |> 
    map_df(~ tibble(
      file_id = .x$id,
      report_id = .x$reportId,
      etag = .x$etag,
      status = .x$status,
      file_name = .x$fileName,
      format = .x$format,
      last_modified = as.numeric(.x$lastModifiedTime) / 1000,  # Convert milliseconds to seconds
      start_date = .x$dateRange$startDate,
      end_date = .x$dateRange$endDate,
      browser_url = .x$urls$browserUrl,
      api_url = .x$urls$apiUrl
    )) |> 
    mutate(last_modified = as.POSIXct(last_modified, 
                                      origin = "1970-01-01", 
                                      tz = "America/Los_Angeles"))
  
  return(files_df)
}

File: py_scripts/get_cm360_files.py
# Getting CM360 Files # 
from googleapiclient.discovery import build
from datetime import datetime, timezone, timedelta

def get_cm360_files(profile_id, report_id, credentials):
    # Build the CM360 service
    cm360 = build("dfareporting", "v4", credentials=credentials)

    # Fetch the files
    response = cm360.reports().files().list(
        profileId=profile_id,
        reportId=report_id
    ).execute()

    files = response.get("items", [])

    # If no files found, return empty list
    if not files:
        print("No files found for this report.")
        return []

    # Convert the file metadata into a list of dicts
    parsed_files = []
    for f in files:
        last_modified_ts = int(f["lastModifiedTime"]) / 1000
        last_modified_dt = datetime.fromtimestamp(last_modified_ts, tz=timezone(timedelta(hours=-8)))  # PST/PDT

        parsed_files.append({
            "file_id": f.get("id"),
            "report_id": f.get("reportId"),
            "etag": f.get("etag"),
            "status": f.get("status"),
            "file_name": f.get("fileName"),
            "format": f.get("format"),
            "last_modified": last_modified_dt,
            "start_date": f.get("dateRange", {}).get("startDate"),
            "end_date": f.get("dateRange", {}).get("endDate"),
            "browser_url": f.get("urls", {}).get("browserUrl"),
            "api_url": f.get("urls", {}).get("apiUrl")
        })

    return parsed_files
  
  # R-safe wrapper
def get_cm360_files_from_r(kwargs):
  return get_cm360_files(
      profile_id=kwargs["profile_id"],
      report_id=kwargs["report_id"],
      credentials=kwargs["credentials"]
    )
